File: telegram/src/main/java/com/merfemor/vkwallwatcher/telegram/TelegramProperties.java
package com.merfemor.vkwallwatcher.telegram;

import org.springframework.boot.context.properties.ConfigurationProperties;
import org.springframework.boot.context.properties.ConstructorBinding;
import org.springframework.validation.annotation.Validated;

import javax.annotation.CheckForNull;
import javax.validation.constraints.NotEmpty;
import javax.validation.constraints.NotNull;

/**
 * @implNote validation not works if implemented in Kotlin
 */
@ConstructorBinding
@ConfigurationProperties(prefix = "telegram")
@Validated
public class TelegramProperties {
    @NotNull
    @NotEmpty
    public final String botUsername;
    @NotNull
    @NotEmpty
    public final String botToken;
    @CheckForNull
    public final String[] permittedUsernames;
    @CheckForNull
    public final String[] adminUsernames;

    private TelegramProperties(String botUsername, String botToken,
                               @CheckForNull String[] permittedUsernames,
                               @CheckForNull String[] adminUsernames) {
        this.botUsername = botUsername;
        this.botToken = botToken;
        this.permittedUsernames = permittedUsernames;
        this.adminUsernames = adminUsernames;
    }
}

File: telegram/src/main/java/com/merfemor/vkwallwatcher/telegram/TelegramProperties.kt
package com.merfemor.vkwallwatcher.telegram

import org.springframework.boot.context.properties.ConfigurationProperties
import org.springframework.boot.context.properties.ConstructorBinding
import org.springframework.validation.annotation.Validated
import javax.validation.constraints.NotEmpty

@ConstructorBinding
@ConfigurationProperties(prefix = "telegram")
@Validated
class TelegramProperties private constructor(
        @NotEmpty val botUsername: String,
        @NotEmpty val botToken: String,
        val permittedUsernames: Array<String>?,
        val adminUsernames: Array<String>?)
